File: library/src/main/java/com/mikepenz/materialdrawer/model/utils/BadgeDrawableBuilder.java
package com.mikepenz.materialdrawer.model.utils;

import android.content.Context;
import android.graphics.drawable.GradientDrawable;
import android.graphics.drawable.StateListDrawable;
import android.support.v7.content.res.AppCompatResources;
import android.util.StateSet;

import com.mikepenz.materialdrawer.holder.BadgeStyle;
import com.mikepenz.materialdrawer.holder.ColorHolder;

/**
 * Created by mikepenz on 02.07.15.
 */
public class BadgeDrawableBuilder {
    private BadgeStyle mStyle;

    public BadgeDrawableBuilder(BadgeStyle style) {
        this.mStyle = style;
    }

    public StateListDrawable build(Context ctx) {
        StateListDrawable stateListDrawable = new StateListDrawable();
        GradientDrawable normal = (GradientDrawable) AppCompatResources.getDrawable(ctx, mStyle.getGradientDrawable());
        GradientDrawable selected = (GradientDrawable) normal.getConstantState().newDrawable().mutate();

        ColorHolder.applyToOrTransparent(mStyle.getColor(), ctx, normal);
        if (mStyle.getColorPressed() == null) {
            ColorHolder.applyToOrTransparent(mStyle.getColor(), ctx, selected);
        } else {
            ColorHolder.applyToOrTransparent(mStyle.getColorPressed(), ctx, selected);
        }

        if (mStyle.getCorners() != null) {
            normal.setCornerRadius(mStyle.getCorners().asPixel(ctx));
            selected.setCornerRadius(mStyle.getCorners().asPixel(ctx));
        }

        stateListDrawable.addState(new int[]{android.R.attr.state_pressed}, selected);
        stateListDrawable.addState(StateSet.WILD_CARD, normal);

        return stateListDrawable;
    }
}

File: library/src/main/java/com/mikepenz/materialdrawer/model/utils/BadgeDrawableBuilder.kt
package com.mikepenz.materialdrawer.model.utils

import android.content.Context
import android.graphics.drawable.GradientDrawable
import android.graphics.drawable.StateListDrawable
import android.util.StateSet
import androidx.appcompat.content.res.AppCompatResources
import com.mikepenz.materialdrawer.holder.BadgeStyle
import com.mikepenz.materialdrawer.holder.ColorHolder

/**
 * Builder to construct the [StateListDrawable] given a [BadgeStyle]
 */
class BadgeDrawableBuilder(private val style: BadgeStyle) {

    /** creates the [StateListDrawable] given the provided [BadgeStyle] */
    fun build(ctx: Context): StateListDrawable {
        val stateListDrawable = StateListDrawable()
        val normal = AppCompatResources.getDrawable(ctx, style.gradientDrawable) as GradientDrawable?
        val selected = normal?.constantState?.newDrawable()?.mutate() as GradientDrawable?

        ColorHolder.applyToOrTransparent(style.color, ctx, normal)
        if (style.colorPressed == null) {
            ColorHolder.applyToOrTransparent(style.color, ctx, selected)
        } else {
            ColorHolder.applyToOrTransparent(style.colorPressed, ctx, selected)
        }

        style.corners?.let {
            normal?.cornerRadius = it.asPixel(ctx).toFloat()
            selected?.cornerRadius = it.asPixel(ctx).toFloat()
        }

        stateListDrawable.addState(intArrayOf(android.R.attr.state_pressed), selected)
        stateListDrawable.addState(StateSet.WILD_CARD, normal)

        return stateListDrawable
    }
}
